sklearn/metrics/_classification.py

```diff
@@ -321,17 +321,26 @@ def confusion_matrix(y_true, y_pred, *, labels=None, sample_weight=None,
                          "'all', None}")
 
     n_labels = labels.size
-    label_to_ind = {y: x for x, y in enumerate(labels)}
-    # convert yt, yp into index
-    y_pred = np.array([label_to_ind.get(x, n_labels + 1) for x in y_pred])
-    y_true = np.array([label_to_ind.get(x, n_labels + 1) for x in y_true])
+    # If labels are not consecutive integers starting from zero, then
+    # yt, yp must be converted into index form
+    need_index_conversion = not (
+        labels.dtype.kind in {'i', 'u', 'b'} and
```

Comment:
Even for small sets, there is a small speed advantage to using sets for `__contains__` checks:

```
def bench_set_membership():
    """
    Q: Is there a speedup to using a set when the set is small?
    A: There seems to be a small benefit.

    Results:
        Timed best=358.996 ns, mean=412.455 ± 67.9 ns for small-set-membership
        Timed best=371.976 ns, mean=411.672 ± 29.0 ns for small-str-membership
        Timed best=362.983 ns, mean=406.722 ± 48.3 ns for small-list-membership

    """
    import timerit
    import random
    ti = timerit.Timerit(300000, bestof=30, verbose=1)
    rng = random.Random(0)

    lhs_candidates = 'biufcmMOSUV'

    for i in range(100000):
        i += 1   # warmup, reduce variance?

    for timer in ti.reset('small-set-membership'):
        lhs = rng.choice(lhs_candidates)
        with timer:
            lhs in {'i', 'u', 'b'}

    for timer in ti.reset('small-str-membership'):
        lhs = rng.choice(lhs_candidates)
        with timer:
            lhs in 'iub'

    for timer in ti.reset('small-list-membership'):
        lhs = rng.choice(lhs_candidates)
        with timer:
            lhs in ['i', 'u', 'b']
```

Also, IMO explicitly separating items is more readable than packing them together in a string.

Reply:
Not that we don't care about the performance of this compared to the actual confusion matrix computation because this is independent of the size of `y_true` and `y_pred` and we only care about the performance when those are big enough to matter.

But I am fine about keeping the code as it is for readability.